### DISCUSSION TITLE
HELP: IGMP monitoring tool

### DISCUSSION THREAD

**Discussion by lars18th** [troglobit/pimd#246] [Q&A]
Hi @troglobit ,

I know that this is not the correct place to make a question not directly related to this tool. However, I'm tired searching for a "simple" solution and I feel you can help me. Sorry if you don't accept this use of the issues section.

My problem is to get a tool that can listen for the IGMP Host Reports in my LAN (that is multicast-aware using the IGMP Snoopig). I want to start/stop a process (than generates the multicast traffic) based on the presence of multicast listeners. It seems to me that such a tool might be easy to find, but it is not. You know about something similar?

Thank you and sorry for this post.

**Comment by troglobit** [comment]
First of all, what you want to do is not how the system is designed or intended so you'll be fighting it all the way if you persist.  I suggest reading up on the protocol first.  The idea is to have senders send all the time (they have something to send) and let the network handle distribution on the LAN and the rest of the network by actively flooding or pruning the multicast.  An IGMP snooping switch, for instance, can either block by default (some vendors do this) or flood by default until they hear a membership report.  If your LAN is in the later category I understand where you're coming from.  If you worry about the spread of your sender's multicast in such networks you could always have the sender also join the group.  This would cause the network to only flood as far as the nearest snooping switch.

To your question, what you're describing is basically a multicast router, like dear old `pimd`, but instead of routing the traffic into your LAN you want to start the local sender.  To achieve this you have to be the router, or the elected IGMP querier on the LAN, otherwise you will not see the host membership reports.

If you don't have an IGMP querier on your LAN you can assume that role, of course, but chances are you do.  In which case you can break multicast filtering on the LAN.  You could try the [multicast router discovery protocol](https://datatracker.ietf.org/doc/html/rfc4286#section-3.2), injecting router advertisements.  I made a [small tool](https://github.com/troglobit/mrdisc/) for that ages ago. But most IGMP snooping switches don't support this.  Another mechanism you can try is to inject PIM HELLO messages in your network, some switches support that, for instance the Linux bridge does.  Both of these methods are intended as a way of gently asking IGMP snooping switches to add the port they hear the messages on as a dynamic multicast router port, and then forward any IGMP membership reports that way.

**Reply by lars18th** [reply]
Hi @troglobit ,

Thank you for answering to my request. And sorry to write it as an issue (I forgot to check the discussion area for this project).

Regarding my use case, I understand how the IGMP (and MLD) works, and what's the objective: pass/block the multicast traffic over the different ports of an switch/router based on active receivers. So it's basically a routing protocol for controlling the L2 traffic of a multicast flow. However, in my professional TV broadcasting background I've seen several times some ON-DEMAND multicast stream generators. They work listening the IGMP messages in the LAN to start/end the multicast stream. Therefore, my idea is to mimic this behaviour to control other IoT services based on the same principle: starting/stoping the service based on the number of listeners (joiners of the multicast group) in the network.

From my point of view, this doesn't needs to be very difficult. And this tool doesn't require to be the master IGMP querier on the LAN. It only needs to subscribe to the specific multicast address, and receive the IGMP reports for this address from the querier. Then, from the complete list of subcribed joiners, it could derive the number of active listeners (i.e. the total number minus one, because the tool is also a listener). And the tool only needs to publish this information in realtime to share it with other processes.

But I'm not sure why this tool doesn't exist yet - am I missing something?

**Reply by troglobit** [reply]
Host membership reports are only forwarded to the active querier (and known multicast routers) on the LAN.  You cannot just subscribe to the group and expect to see what other hosts do.  In the past, when the first IGMP standard was written, it was written to accommodate hubbed networks, but that does not apply anymore in a switched networks.  So there is no such "simple" tool because it's not a simple problem to solve.

**Reply by lars18th** [reply]
> So there is no such "simple" tool because it's not a simple problem to solve.

Ok. Then this tool requires to be developed. One question now: Any idea to accomplish with this behaviour? I feel you have a lot of expertice with the IGMP standard and can suggest something that will have sense. Please comment.

**Reply by troglobit** [reply]
This seems to be a you problem, I have given you all the help and time I can spare.
